### DISCUSSION TITLE
Server logs to Postgresql

### DISCUSSION THREAD

**Discussion by houshmandX** [nextgenhealthcare/connect#5186] [Q&A] [answered]
Hi, 

I found the information in log4j about how to write server logs into the database but pictured some issues. Is there anyone has done this before? sending server logs to Postgres?  
https://www.tutorialspoint.com/log4j/log4j_logging_database.htm
https://forums.mirthproject.io/forum/mirth-connect/support/11487-log-info

**Comment by jonbartels** [comment]
A few things to be aware of:
* MC uses Log4j 1.x (https://github.com/nextgenhealthcare/connect/discussions/4892) - When you're reading about Log4J configuration, remember to make sure you're reading the 1.x documentation and not the 2.x documentation
* https://github.com/odoodo/Mirth-MetaAppender is a great custom logger for MC. It does not log to the database but it is a good reference for how logging in MC works and some powerful things you can do with it

I would also be curious about what happens in Log4J if the DB connection is lost or the DB runs out of space. What happens to the logs and the application?

What problem are you trying to solve? Do you have large log files (logging less data might be better, log4j and logrotate also have compression options)? Do you need to preserve log files for a long time (is a DB a good choice or a tool like Graylog or some other log archiver)?

**Reply by houshmandX** [reply]
Hi, thanks for the response. I need to store/back-up server logs(where if mirth crashes and some errors happen) daily and save them into DB(then transfer to the s3 bucket and empty the DB) or send the daily logs to s3. In this case, if something happens I can go back and track the issues in the logs. Currently, server logs are in 20 folders, which get overridden when all pass the capacity(increasing the numbers or file capacity won't help since we cannot be guaranteed to capture all of the daily logs). 
in a short sentence: I want to capture all daily logs and save them into the database or AWS s3 bucket.

**Reply by jonbartels** [reply]
I say add the database appender and try it out! 

AWS S3 would require something like copying the rotated log files or a database. My understanding is that S3 is not well suited to streaming files (I could be wrong though!).

**Reply by siddharthsharma9258** [reply]
..Then you need to log the log calls to the DB. Sounds like log recursion! :)

Sorry couldn't resist the joke.

**Reply by houshmandX** [reply]
adding below along previus appenders: 
```
# Define the DB appender
log4j.appender.DB=org.apache.log4j.jdbc.JDBCAppender

# Set JDBC URL

log4j.appender.DB.URL=jdbc:postgresql://XXX

# Set Database Driver
log4j.appender.DB.driver=org.postgresql.Driver
# Set database user name and password
log4j.appender.DB.user=XX
log4j.appender.DB.password=XX

# Set the SQL statement to be executed.
log4j.appender.DB.sql=INSERT INTO LOGS VALUES('%x','%d','%C','%p','%m')

# Define the layout for file appender
log4j.appender.DB.layout=org.apache.log4j.PatternLayout
```

also created a LOGS table in Postgres 


no result!

what am I missing here?

**Reply by jonbartels** [reply]
Ideas:
* Does `mirth.log` show any errors?
* Does your DB show connection attempts with the logging user to the logging DB? Any errors in the DB logs?
* Did you add `DB` to `log4j.rootLogger = ERROR,stdout,fout` ? (this sets which log appenders are actually used or enabled for the root logger)
* Start MC with `./mcserver`. It launches MC as a command line app, it will often record more info to the console when it has problems with logging or problems that happen before the logger starts
* Does the JDBCAppender JAR exist in the MC classpath? Either `server-lib` or `custom-lib` with the `includecustomlib` property set
* Add `-Dlog4j.debug=true` to `mcserver.vmoptions`, this will cause log4j itself to print out log4js own information when you start MC with `./mcserver`

**Reply by houshmandX** [reply]
it seems like the issue is the org.postgresql.Driver which leads to the following error:
```
log4j:ERROR Failed to load driver
java.lang.ClassNotFoundException: org.postgresql.Driver
```

I have checked that the Postgres jar file exists in the database folder, and the server can connect to Postgres and write events, etc.

how to fix the issue of log4j with the driver?

**Reply by ab-20-m** [reply]
The regular server-lib directory is not loaded into the classpath until after logging is initialized. You will need to copy the Postgres JDBC JAR into the server-launcher-lib folder or use the exe4j .vmoptions -classpath/p setting to prepend the JDBC driver at the start of the classpath for Mirth. A symlink to the JAR in the server-launcher-lib may also work but I seem to recall Java can have issues loading JARs that way.

The code where the classpath is built up before logging initialization is in [MirthLauncher.java L54-L61](https://github.com/nextgenhealthcare/connect/blob/492256368773a427a00c9d1b1de8240ea376c1a9/server/src/com/mirth/connect/server/launcher/MirthLauncher.java#L54-L61).

**Reply by houshmandX** [reply]
here is where I create my table:
`CREATE TABLE LOGS
   (USER_ID VARCHAR(20)    NOT NULL,
    DATED   DATE           NOT NULL,
    LOGGER  VARCHAR(50)    NOT NULL,
    LEVEL   VARCHAR(10)    NOT NULL,
    MESSAGE VARCHAR(1000)  NOT NULL
   );`

here is where I insert to table:
```
# Set the SQL statement to be executed.
log4j.appender.DB.sql=INSERT INTO LOGS VALUES('%x','%d','%C','%p','%m')
# Define the layout for file appender
log4j.appender.DB.layout=org.apache.log4j.PatternLayout
```
<img width="759" alt="Screen Shot 2022-05-16 at 11 19 00 AM" src="https://user-images.githubusercontent.com/56044836/168657327-82eaa6de-84af-4149-b609-d9cf5086eac8.png">

and here is the error I get:

`ERROR 2022-05-16 11:20:49,953 [Main Server Thread] Server: log4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute`

how to fix the issue?

**Reply by jonbartels** [reply]
> ERROR 2022-05-16 11:20:49,953 [Main Server Thread] Server: log4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute

Is there any more context to this error? Is there a stack trace from MC in the console or in `mirth.log`?

Any errors in your `postgres.log` for Postgres?

**Reply by houshmandX** [reply]
nothing beside bunch of same error:

`ERROR 2022-05-16 14:16:21,892 [Main Server Thread] Server: log4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute sqllog4j:ERROR Failed to excute `

**Reply by houshmandX** [reply]
FYI. I could able to fix the issues and write them into database. thanks for all help and hint.

**Comment by houshmandX** [comment, marked as answer]
here is the solution from my end:

1. for creating a table increase the varchar for logger and change the d(data type), so create the table as below:
 
 `CREATE TABLE LOGS (USER_ID VARCHAR(20) NOT NULL, DATED DATE NOT NULL, LOGGER VARCHAR(100) NOT NULL, LEVEL VARCHAR(10) NOT NULL, MESSAGE VARCHAR(1000) NOT NULL );`

2. add/copy postgres_jdbc.jar from the `database` folder in MC, and past/place it in `server-launcher-lib` folder in MC(mirth connect).
3. change date format '%d' to '%d{yyyy-MM-dd}'

here is my final log4j.properties that saves server logs to Postgres db:

```
# sample properties to initialize log4j
log4j.rootLogger = ERROR,stdout,fout,DB

# stdout appender
log4j.appender.stdout = org.apache.log4j.ConsoleAppender
log4j.appender.stdout.layout = org.apache.log4j.PatternLayout
log4j.appender.stdout.layout.ConversionPattern = %-5p %d [%t] %c: %m%n


# file appender
dir.logs = C:\\Program Files\\Mirth Connect\\logs
log4j.appender.fout = org.apache.log4j.RollingFileAppender
log4j.appender.fout.File = ${dir.logs}/mirth.log
log4j.appender.fout.MaxFileSize = 500KB
log4j.appender.fout.MaxBackupIndex = 20
log4j.appender.fout.layout = org.apache.log4j.PatternLayout
log4j.appender.fout.layout.ConversionPattern = %-5p %d [%t] %c: %m%n

# Define the DB appender
log4j.appender.DB=org.apache.log4j.jdbc.JDBCAppender

# Set JDBC URL
log4j.appender.DB.URL=jdbc:postgresql://XXX:5432/DBNAME

# Set Database Driver
log4j.appender.DB.driver=org.postgresql.Driver
log4j.appender.DB.user=USERNAME
log4j.appender.DB.password=PASSWORD

# Set the SQL statement to be executed.
log4j.appender.DB.sql=INSERT INTO LOGS VALUES('%x','%d{yyyy-MM-dd}','%C','%p','%m')
# Define the layout for file appender
log4j.appender.DB.layout=org.apache.log4j.PatternLayout


# splash screen
log4j.logger.com.mirth.connect.server.Mirth = INFO

# Mirth Connect server logging
log4j.logger.com.mirth.connect.server.controllers.DonkeyEngineController = INFO
log4j.logger.com.mirth.connect.donkey.server.channel.RecoveryTask = INFO
log4j.logger.com.mirth.connect.connectors.file.FileReceiver = WARN

# Mirth Connect channel logging
log4j.logger.transformer = INFO
log4j.logger.preprocessor = INFO
log4j.logger.postprocessor = INFO
log4j.logger.deploy = INFO
log4j.logger.undeploy = INFO
log4j.logger.filter = INFO
log4j.logger.db-connector = INFO
log4j.logger.js-connector = INFO
log4j.logger.attachment = INFO
log4j.logger.batch = INFO
log4j.logger.response = INFO

# SQL Logging
log4j.logger.java.sql = ERROR
```

**Reply by 1jon11** [reply]
I just tried this and not working to me. I'm using Mirth Connect 4.4.2. 

I edited log4j2.properties , on /conf directory

Any idea?

**Reply by pacmano1** [reply]
Mirth 4.4.2 uses log4j 2 not 1.   Also, create a new discussion thread.  Also don't post "it does not work for me" without thorough details.
